### PR TITLE
Add `when` context manager to group common constraints in packages

### PR DESCRIPTION
**Pull request by alalazo**
This PR adds a context manager that permit to group the common part of a `when=` argument and add that to the context:
```python
class Gcc(AutotoolsPackage):
    with when('+nvptx'):
        depends_on('cuda')
        conflicts('@:6', msg='NVPTX only supported in gcc 7 and above')
        conflicts('languages=ada')
        conflicts('languages=brig')
        conflicts('languages=go')
```
The above snippet is equivalent to:
```python
class Gcc(AutotoolsPackage):
    depends_on('cuda', when='+nvptx')
    conflicts('@:6', when='+nvptx', msg='NVPTX only supported in gcc 7 and above')
    conflicts('languages=ada', when='+nvptx')
    conflicts('languages=brig', when='+nvptx')
    conflicts('languages=go', when='+nvptx')
```
which needs a repetition of the `when='+nvptx'` argument. The context manager might help improving readability and permits to group together directives related to the same semantic aspect (e.g. all the directives needed to model the behavior of `gcc` when `+nvptx` is active). 

Modifications:

- [x] Added a `when` context manager to be used with package directives
- [x] Add unit tests and documentation for the new feature
- [x] Modified `cp2k` and `gcc` to show the use of the context manager

### CODE REVIEW

**Comment by alalazo** [comment]
@tgamblin I know we discussed calling the context manager `when`, but it turns out we already have a decorator in scope with that name:
```python
@when('@1.0')
def config_args(self):
    pass
```

**Comment by haampie** [comment]
This looks neat :)

**Comment by tgamblin** [comment]
> I know we discussed calling the context manager when, but it turns out we already have a decorator in scope with that name

@alalazo does it matter?  the decorator is currently callable (which makes it a decorator) but why can't it also be a context manager?  I don't think overloading is bad here -- it's clear from context (ha) what's meant.

**Comment by tgamblin** [comment]
Also, both the decorator and the context manager do a similar thing and take similar args.

**Comment by tgamblin** [comment]
Hoping that if @adamjstewart sees this it will trigger a flood of OCD changes.

**Comment by adamjstewart** [comment]
Don't tell my manager I'm working on Spack during my internship 😜

**Comment by alalazo** [comment]
> I don't think overloading is bad here -- it's clear from context (ha) what's meant

Since you asked me to abuse Python, I don't see why I should say "no" :smirk: Now "when" is both a context manager AND a decorator.

**Comment by tgamblin** [comment]
> context manager AND a decorator

Eeeeexcellent.  A `contanorator`?  We clearly need a term for this.

File: lib/spack/docs/packaging_guide.rst [hunk 1]
```diff
@@ -2775,6 +2775,57 @@ packages be built with MVAPICH and GCC.
 
 See the :ref:`concretization-preferences` section for more details.
 
+
+.. _group_when_spec:
+
+------------------------------------
+Group ``when=`` constraints together
```
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]: Can we call this "Common `when=` constraints"?

File: lib/spack/docs/packaging_guide.rst [hunk 2]
```diff
@@ -2775,6 +2775,57 @@ packages be built with MVAPICH and GCC.
...
+------------------------------------
+
+In case a package needs many directives to share the whole ``when=``
+argument, or just part of it, Spack permits to group the common part
```
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]: Instead of "permits to" can you say "allows you to"? (here and elsewhere)

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Comment by tgamblin** [comment]
@alalazo just a couple really small requests

File: lib/spack/docs/packaging_guide.rst (hunk 1 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: See e0e1d74

File: lib/spack/docs/packaging_guide.rst (hunk 2 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: See e0e1d74

**Review by tgamblin — APPROVED** [review]

**Pull request merged**